src/pip/_internal/index.py

```diff
@@ -1333,9 +1336,17 @@ def _clean_link(url):
         path = urllib_request.pathname2url(
             urllib_request.url2pathname(result.path))
     else:
-        # In addition to the `/` character we protect `@` so that
-        # revision strings in VCS URLs are properly parsed.
-        path = urllib_parse.quote(urllib_parse.unquote(result.path), safe="/@")
+        # In addition to the `/` character we protect `@`
+        # (as well as theirs %-escapes) so that revision
```

Comment:
I don't think the %-escapes are for VCS URLs to be properly parsed as much as avoiding side-effects of our preserving them. Like:

* We preserve `@` as-is (in addition to the default `/`) so that VCS URLs are properly parsed
* We preserve their %-escapes so they are not inadvertently converted to the literal characters

Reply:
Do you suggest to update the comment? WDYT maybe it's better to set this comment above the `_safe_chars_re`?

Reply:
Maybe we could say: We quote the path for convenience, but VCS URL paths have a plain `@` which we should preserve. To avoid double-quoting we unquote then quote, but extract any manually quoted `@` and `/` beforehand so they are not accidentally preserved.